Run clang-tidy against ESP32

oxan opened the pull request:
# What does this implement/fix? 

First steps to run clang-tidy against the ESP32 code.

There's still a bunch of errors that need to be fixed. It might also be a bit overkill to run clang-tidy twice on the full codebase in CI, but I'm not sure how to avoid that yet.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

oxan committed:
Allow specifying environment for clang-tidy to run against

oxan committed:
Fix a bunch of clang-tidy warnings on the ESP32

oxan committed:
Add flags for clang-tidy to succeed for the ESP32

oxan committed:
Add ESP32 clang-tidy to CI

oxan committed:
Fix a bunch more clang-tidy warnings on the ESP32

OttoWinter commented:
As you said running clang-tidy twice for every file will slow down CI (fixed number of workers for whole org)

Maybe this works:
1. The majority of platform-dependent code is for ESP32 p, so I think we can use ESP32 by-default, and ESP8266 only in certain cases
2. Only run ESP8266 checks against files that contain `ARDUINO_ARCH_ESP8266` (or `ESP8266`), that should work quite well since those ifdefs need to be there anyway so that tidy doesn't complain when running in ESP32 mode

oxan committed:
Don't run clang-tidy twice for every file in CI

oxan committed:
Fix a bunch more ESP32 clang-tidy issues

oxan committed:
Use lower case names for static variables

oxan commented:
>   1. The majority of platform-dependent code is for ESP32 p, so I think we can use ESP32 by-default, and ESP8266 only in certain cases
>   2. Only run ESP8266 checks against files that contain `ARDUINO_ARCH_ESP8266` (or `ESP8266`), that should work quite well since those ifdefs need to be there anyway so that tidy doesn't complain when running in ESP32 mode

I've implemented this now, and also fixed or suppressed the remaining issues on the ESP32, so this can be reviewed/merged now.

jesserockz committed:
Merge branch 'dev' into tidy/esp32

oxan committed:
Remove unnecessary USE_ETHERNET build flag

oxan committed:
Fix clang-tidy warnings in airthings integration

oxan committed:
Include USE_ADC_SENSOR_VCC only for ESP8266

oxan committed:
Fix a couple of new clang-tidy warnings

oxan committed:
Fix memory ownership warnings on ESP32

oxan committed:
Merge branch 'dev' into tidy/esp32

oxan committed:
Fix missing LogString conversion

oxan commented on esphome/components/nextion/nextion_upload.cpp:
```diff
@@ -25,15 +25,10 @@ int Nextion::upload_by_chunks_(HTTPClient *http, int range_start) {
 #ifndef CLANG_TIDY
   http->setFollowRedirects(HTTPC_STRICT_FOLLOW_REDIRECTS);
   http->setRedirectLimit(3);
-  begin_status = http->begin(*this->get_wifi_client_(), this->tft_url_.c_str());
```
I've deleted the `begin` calls here, as they seem useless since the first thing that happens in the `while` loop below is to call `begin` again. However, I don't have the hardware so haven't been able to test this.

CC @SenexCrenshaw

kbx81 replied:
I have the hardware and can test it out later today to confirm. 😄

oxan commented on esphome/components/uart/uart_esp32.cpp:
```diff
@@ -99,7 +99,7 @@ void UARTComponent::dump_config() {
   ESP_LOGCONFIG(TAG, "  Baud Rate: %u baud", this->baud_rate_);
   ESP_LOGCONFIG(TAG, "  Data Bits: %u", this->data_bits_);
-  ESP_LOGCONFIG(TAG, "  Parity: %s", parity_to_str(this->parity_));
+  ESP_LOGCONFIG(TAG, "  Parity: %s", LOG_STR_ARG(parity_to_str(this->parity_)));
```
This one is harmless as the macro resolves to a runtime no-op on ESP32, but it breaks typing of course. Curious that gcc didn't complain about this.

oxan committed:
Initialize variable to make clang-tidy happy

oxan committed:
Merge branch 'dev' into tidy/esp32

oxan commented:
Remaining errors are all due to memory ownership issues in the BLE components. That needs to be fixed, but I have no experience with that code, and converting it to unique_ptr looks to be complicated...

OttoWinter reviewed (approved):
LGTM, thanks very much!!

OttoWinter committed:
Fixes for dev activated owning memory check

OttoWinter committed:
Merge remote-tracking branch 'origin/dev' into tidy/esp32

OttoWinter committed:
Format

OttoWinter committed:
Fix
